![venomancer_logo](/venomancer_logo.png)

Venomancer
==========
<!---[![Latest GitHub release](https://img.shields.io/github/release/pedronachtigall/Venomancer.svg)](https://github.com/pedronachtigall/Venomancer/releases/latest) -->
<!---[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3403273.svg)](https://doi.org/10.5281/zenodo.3403273) -->
<!---[![Published in Genome Biology](https://img.shields.io/badge/published%20in-Genome%20Biology-blue.svg)](https://doi.org/10.1101/gr.214270.116) -->

**Venomancer** is a computational tool designed to detect and annotate toxin genes in transcriptome assembly.

Getting Started
=================

# Installation

Download the master folder and follow the steps below:

```
unzip Venomancer-master.zip
export PATH=$PATH:path/to/Venomancer-master/bin/
```

# Requirements

- [Python3](https://www.python.org/) and [Biopython](https://biopython.org/wiki/Download)
  - ```apt-get install python3-biopython```
- [Perl](https://www.perl.org/), [Bioperl](https://bioperl.org/) and [MCE](https://metacpan.org/release/MCE) (libmce-perl)
  - ```apt-get install bioperl libmce-perl```
- [CodAn](https://github.com/pedronachtigall/CodAn/)
- [NCBI-BLAST](https://www.ncbi.nlm.nih.gov/books/NBK279671/) (v2.9.0 or above)
- [SignalP-4.1](http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+4.1)


# Usage

```

```

# Inputs


# Outputs


Reference
=========

If you use or discuss **Venomancer**, please cite:

Nachtigall et al., under review

License
=======

<!---[GNU GPLv3](https://www.gnu.org/licenses/gpl-3.0.html) -->

Contact
=======

To report bugs, to ask for help and to give any feedback, please contact **Pedro G. Nachtigall**: pedronachtigall@gmail.com

